cp1616/src/cp1616_io_device.cpp

```diff
@@ -0,0 +1,798 @@
+/*********************************************************************************************//**
+* @file cp1616_io_device.cpp
+*
+* cp1616_io_device class
+*
+* Copyright {2015} {Frantisek Durovsky}
+*
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+
+*      http://www.apache.org/licenses/LICENSE-2.0
+
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+
+* *********************************************************************************************/
+#ifndef CP1616_IO_DEVICE_CPP
+#define CP1616_IO_DEVICE_CPP
+
+#include <cp1616/cp1616_io_device.h>
+#include <cp1616/cp1616_io_device_callbacks.h>
+
+//Define and initialize device_instance_ to zero value;
+Cp1616IODevice *Cp1616IODevice::device_instance_ = 0;
+
+Cp1616IODevice* Cp1616IODevice::getDeviceInstance()
+{
+  if( !device_instance_ )
+  {
+    device_instance_ = new Cp1616IODevice();
+  }
+  return device_instance_;
+}
+
+Cp1616IODevice::Cp1616IODevice():
+    cp_id_(1),
+    cp_handle_(0),
+    cp_session_key_(0),
+    cp_ar_number_(0),
+    ar_info_ind_flag_(0),
+    prm_end_ind_flag_(0),
+    indata_ind_flag_(0),
+    offline_ind_flag_(0),
+    p_device_data_(NULL)
+{
+
+    p_device_data_ = new device_data_t [DEVICE_DATA_ENTRIES * sizeof(device_data_t)];
+
+    //Initialize device_data_ according to STEP project setup
```

Comment:
Of course not, this is just for initial testing. I wanted to discuss the initialization as a next step. I intend to use .yaml config file to hold module parameters from step7 config, (slot, subslot, modID, subslotID, + ROS side parameters - like topic name, type of data variable and maybe rate) to load them on rosparam server to make them available for the application.  **ros_canopen** utilizes the same pricniple. 

There will be a "parser" (is parser a right name for such a purpose?) that will fill DeviceData structure according to yaml config in constructor.